art/attacks/fast_gradient.py

```diff
@@ -150,14 +162,22 @@ def generate(self, x, **kwargs):
         # Return adversarial examples computed with minimal perturbation if option is active
         if 'minimal' in params_cpy and params_cpy[str('minimal')]:
             logger.info('Performing minimal perturbation FGM.')
-            x_adv = self._minimal_perturbation(x, y, **params_cpy)
+            adv_x_best = self._minimal_perturbation(x, y, **params_cpy)
+            rate_best = 100 * compute_success(self.classifier, x, y, adv_x_best, self.targeted)
         else:
-            x_adv = self._compute(x, y, self.eps, self.eps, self.random_init)
+            adv_x_best = None
+            rate_best = 0.0
 
-        logger.info('Success rate of FGM attack: %.2f%%',
-                    100 * compute_success(self.classifier, x, y, x_adv, self.targeted))
+            for i_random_init in range(max(1, self.num_random_init)):
+                adv_x = self._compute(x, y, self.eps, self.eps, self.num_random_init > 0)
+                rate = 100 * compute_success(self.classifier, x, y, adv_x, self.targeted)
+                if rate > rate_best or adv_x_best is None:
+                    rate_best = rate
+                    adv_x_best = adv_x.copy()
```

Comment:
Do you need to copy here? Or just assignment? Same goes for the other places you are using this construct.

Reply:
Resolved with 0265858d96c93eb154e42213b14a4855274e8c01